src/linear_algebra/free_module/basic.lean

```diff
@@ -113,16 +105,27 @@ of_equiv e
 
 variables (R M N)
 
-instance {ι : Type v} : module.free R (ι →₀ R) :=
-of_basis (basis.of_repr (linear_equiv.refl _ _))
-
-instance {ι : Type v} [fintype ι] : module.free R (ι → R) :=
-of_equiv (basis.of_repr $ linear_equiv.refl _ _).equiv_fun
+/-- The module structure provided by `semiring.to_module` is free. -/
+instance self : module.free R R := of_basis (basis.singleton unit R)
 
 instance prod [module.free R N] : module.free R (M × N) :=
 of_basis $ (choose_basis R M).prod (choose_basis R N)
 
-instance self : module.free R R := of_basis $ basis.singleton unit R
+/-- The product of finitely many free modules is free. -/
+instance pi (M : ι → Type*) [finite ι] [Π (i : ι), add_comm_monoid (M i)]
+  [Π (i : ι), module R (M i)] [Π (i : ι), module.free R (M i)] : module.free R (Π i, M i) :=
+let ⟨_⟩ := nonempty_fintype ι in by exactI (of_basis $ pi.basis $ λ i, choose_basis R (M i))
+
+/-- The module of finite matrices is free. -/
+instance matrix {m n : Type*} [finite m] [finite n] : module.free R (matrix m n M) :=
```

Comment:
Is this still needed?

Reply:
Yes, typeclass search doesn't unfold `matrix` (which is by design)